### PR TITLE
MF-379 - Log Level Option

### PR DESCRIPTION
**Pull request by mfinley3**
### What does this do?
Adds a log level option for each component of mainflux so users can set what logs they want.

### Which issue(s) does this PR fix/relate to?
#379 

### List any changes that modify/break current functionality
Nothing should break/modify default functionality.

### Have you included tests for your changes?
Unit tests for the log level have been added

### Did you document any new/modified functionality?
Yes, readmes have been updated with the new log level env option.

### Notes
If any bad input is given or level is left unselected, level defaults to info. 

Regarding the size of this PR, let me know if you would me to break this up in some way since there were two vendor files added

Signed-off-by: mfinley3 <Michael.Finley@target.com>

### CODE REVIEW

**Comment by chombium** [comment]
@mfinley3 is there any specific reason why you've set 'info' as the default log level?
Usually 'error' is set as default level so that  in production the logs are not flooded with messages which would show that the system is working properly and you keep the config files smaller. Using something other then 'error' in production would mean that you are examining something in the system.

I would suggest that you to change the default level to 'error'.

@mainflux/contributors any thoughts on this?

**Comment by drasko** [comment]
@mfinley3 I agree with @chombium - `error` should be the default level.

File: cmd/cassandra-reader/main.go
Comment (drasko) [inline cmd/cassandra-reader/main.go]: Please change to `error`

File: cmd/cassandra-writer/main.go
Comment (drasko) [inline cmd/cassandra-writer/main.go]: Should be `error`.

File: cmd/http/main.go
Comment (drasko) [inline cmd/http/main.go]: Same here.

File: cmd/influxdb-reader/main.go
Comment (drasko) [inline cmd/influxdb-reader/main.go]: Same here.

File: cmd/influxdb-writer/main.go
Comment (drasko) [inline cmd/influxdb-writer/main.go]: And here.

File: cmd/mongodb-reader/main.go
Comment (drasko) [inline cmd/mongodb-reader/main.go]: And here.

File: cmd/things/main.go
Comment (drasko) [inline cmd/things/main.go]: And here.

File: cmd/users/main.go
Comment (drasko) [inline cmd/users/main.go]: And here

File: cmd/ws/main.go
Comment (drasko) [inline cmd/ws/main.go]: Same here.

File: ws/README.md
Comment (drasko) [inline ws/README.md]: Default log level is error.

File: writers/mongodb/messages_test.go
Comment (drasko) [inline writers/mongodb/messages_test.go]: Strign consts should be defined at the top of the file.

File: users/README.md
Comment (drasko) [inline users/README.md]: Default -> error

File: cmd/normalizer/main.go
Comment (drasko) [inline cmd/normalizer/main.go]: default to error

File: http/README.md
Comment (drasko) [inline http/README.md]: Default -> err

**Review by drasko — CHANGES_REQUESTED** [review]
Very good code here @mfinley3 - just fix default logging level and we are there as long as I am concerned.

File: writers/mongodb/messages_test.go
Comment (mfinley3) [reply writers/mongodb/messages_test.go]: Good catch here, I have made this change to all files similar to this as well!
Ex: ws/api/transport_test.go

**Comment by mfinley3** [comment]
Default level updated to error :)
Let me know if there is anything else!

**Review by drasko — DISMISSED** [review]
LGTM

**Review by manuio — DISMISSED** [review]
LGTM!

**Comment by chombium** [comment]
well done @mfinley3 
LGTM!

File: logger/logger.go
Comment (anovakovic01) [inline logger/logger.go]: I would suggest adding enumeration instead of plain strings for log levels. Also, add method for casting string to enumeration values so that it can be used from `main.go`.

**Comment by drasko** [comment]
Thanks for review @chombium. Let's hold the merge until @anovakovic01 approve, he mentioned quickly that he would prefer enum instead of string map for log level. I'll let him comment.

File: logger/logger.go
Comment (anovakovic01) [inline logger/logger.go]: Please return error if invalid log level is provided.

File: logger/logger.go
Comment (anovakovic01) [inline logger/logger.go]: Use this enumeration instead of string.

**Review by anovakovic01 — CHANGES_REQUESTED** [review]

**Comment by drasko** [comment]
Exaple for enumeration for log level and string representation: https://github.com/uber-go/zap/blob/f4243df2af592a2e0ccdccb77e1ad44f78d735e5/zapcore/level.go#L34

File: logger/logger.go
Comment (mfinley3) [reply logger/logger.go]: I'm struggling a bit with this one. 
I removed the enumeration in [mainflux/level.go](https://github.com/mainflux/mainflux/pull/402/files#diff-a8b26aac2e521bc9f1a99413d2713efaL1) in favor of [vendor/level.go](https://github.com/mainflux/mainflux/blob/b07c6ae26499758753de2b223fab747610dceef9/vendor/github.com/go-kit/kit/log/level/level.go#L178-L205).
It seemed go-kit/log had already provided this functionaily. 

You're able to do `level.InfoValue().String()` similarly to previously, `log.Info.String()` in order to use from `main.go`. 

Admittedly I do like the look of  `log.Info.String()` more, but this removes what would be duplicate code.

Regarding the map I have created, it is being used as a way to convert what would be the env-var logLevel (type string) into a function. Adding a level filter requires a function be passed in [vendor/level.go](https://github.com/mainflux/mainflux/blob/b07c6ae26499758753de2b223fab747610dceef9/vendor/github.com/go-kit/kit/log/level/level.go#L75). So this is how I was mapping that user input string into the function the vendor package needs. Ex: `l = level.NewFilter(l, level.AllowError())`

I can change it to:
```
var logLevels = map[string]level.Option{
	level.DebugValue().String(): level.AllowDebug(),
	level.InfoValue().String():  level.AllowInfo(),
	level.WarnValue().String():  level.AllowWarn(),
	level.ErrorValue().String(): level.AllowError(),
}
```
If that would be prefered.

@anovakovic01 / @drasko  Please let me know your thoughts and please let me know if I misunderstood what was being suggested here!

File: logger/logger.go
Comment (mfinley3) [reply logger/logger.go]: Please take a look at my above comment. I have New accept a string here because that will be coming directly from env-vars like [here](https://github.com/mainflux/mainflux/blob/b07c6ae26499758753de2b223fab747610dceef9/cmd/http/main.go#L51-L55)

Comment (drasko) [reply logger/logger.go]: You can use function similar to [this one](https://github.com/uber-go/zap/blob/f4243df2af592a2e0ccdccb77e1ad44f78d735e5/zapcore/level.go#L116) to unmarshal text into a level struct. But what @anovakovic01 refered to here is that he wants `New()` to take `logLevel` enum as a param, not a string.

File: logger/logger.go
Comment (drasko) [reply logger/logger.go]: Idea is to create our levels, independent of go-kit logger, i.e. create an abstraction/wrapper there, so that we can replace go-kit logger in the future - for example with [zap](https://github.com/uber-go/zap).

Comment (mfinley3) [reply logger/logger.go]: Ah ok that makes sense, I will work towards that approach and remove the dependency on go-kits levels

Comment (anovakovic01) [reply logger/logger.go]: Thanks @mfinley3.

**Review by dborovcanin — CHANGES_REQUESTED** [review]

File: cmd/things/main.go
Comment (mfinley3) [reply cmd/things/main.go]: Good catch @dusanb94, I've updated it to use the cacheDB parameter in the lines below!

File: logger/logger.go
Comment (mfinley3) [reply logger/logger.go]: @anovakovic01 I've made some changes and it now uses the level wrapper and steers clear of go-kit/level. Please let me know your thoughts on the changes!

File: logger/logger.go
Comment (mfinley3) [reply logger/logger.go]: I've started returning an error and no longer default it to error if there is bad input.

Please let me know your thoughts on how I chose to handle it for now.
https://github.com/mainflux/mainflux/blob/2ffb2bd3ba384c7e9451fdb30862d9cbb0d0511f/cmd/normalizer/main.go#L95-L100
To be honest I'm not the biggest fan of the above; it got a little funky since there is no logger created yet. 

So I think a few options are have each component create a logger and then update it with the appropriate logging level, default if it is bad input, or the above here. 
Thoughts?

**Comment by mfinley3** [comment]
I just realized there are some conflicts, I will work on getting those fixed up in the morning.

File: logger/logger_test.go
Comment (anovakovic01) [inline logger/logger_test.go]: You should move these into `TestDebug`. Just create new logger instance in for loop and set log level from test case. Same for other table maps in this function.

**Review by anovakovic01 — CHANGES_REQUESTED** [review]
Almost there. Good job @mfinley3

File: cmd/cassandra-reader/main.go
Comment (mfinley3) [reply cmd/cassandra-reader/main.go]: Sounds good!

**Comment by mfinley3** [comment]
I've rebased and addressed the conflicts. 
The conflicts were primarily around: [cmd/influxdb-writer/main.go](https://github.com/mainflux/mainflux/pull/402/files#diff-10dba4d266dfb116ff41d69e3950fab7)

Please let me know if there is anything else!

File: cmd/cassandra-reader/main.go
Comment (drasko) [inline cmd/cassandra-reader/main.go]: Instead of unmarshalling the text in each of the `main`s maybe it would be finally better to just pass the string and unmarshall stuff in `logger.New()` - that can do the validation (by trying to unmarshal text) and return error if it fails.

@anovakovic01 @dusanb94 - opinions?

@mfinley3 sorry about this comment, I know we told you to use lvels struct, but for me it was a bit hard to see without implementation - this patch is pretty big.

File: cmd/influxdb-writer/main.go
Comment (drasko) [inline cmd/influxdb-writer/main.go]: No need for this function `unmarshalInfluxDBSettings()` IMHO. I think it hurts readability.

I would rather go with:
```
        batchTimeout, err = strconv.Atoi(cfgBatchTimeout)
	if err != nil {
		// log error
                // os.Exit(1)
	}
 	batchSize, err = strconv.Atoi(cfgBatchSize)
	if err != nil {
		// log error
                // os.Exit(1)
	}
```
directly here.

@anovakovic01 @dusanb94 opinions?

**Review by drasko — COMMENTED** [review]
Just these two minor things from my side, but this is now neatpicking...

File: http/README.md
Comment (dborovcanin) [inline http/README.md:16]: I think that there is a code block explaining how to run a service just under these tables. Should look something like:
```
MF_THINGS_URL=[Things service URL] MF_NATS_URL=[NATS instance URL] MF_HTTP_ADAPTER_PORT=[Service HTTP port] $GOBIN/mainflux-http
```
Please, add this env variable there, as well. Please do this in the other README.md files, as well.

**Review by dborovcanin — COMMENTED** [review]
I agree with @drasko about `unmarshalInfluxDBSettings()` and `logger.New()`. I think what he proposed is cleaner if nothing else. Also, I've found **not handling zero-values for timeout and batch size in InfluxDB writer really dangerous**. It's my bad, so I'll send a separate NOISSUE pull request for that. Great work @mfinley3!

File: http/README.md
Comment (mfinley3) [reply http/README.md:16]: ~Ah yep, I missed this, I will get that cleaned up in the next commit!~

Should be added to all of the readmes now!

File: cmd/influxdb-writer/main.go
Comment (dborovcanin) [reply cmd/influxdb-writer/main.go]: @drasko I agree. @mfinley3 However, if you want to keep this method, I suggest you rename to something like `loadInfluxDBConfig`. I also suggest not to return an error but to simply log and kill in case of wrong env variable values.

Comment (mfinley3) [reply cmd/influxdb-writer/main.go]: I agree as well, I think the function hurts readability. I will try somethings and get this looking better

@dusanb94 I will type up both solutions and see which one looks better. I think I am leaning towards simply removing the function right now though

File: cmd/cassandra-reader/main.go
Comment (mfinley3) [reply cmd/cassandra-reader/main.go]: I liked string initially because of the cleanliness but I do see why passing in an enum for levels makes sense. So, I'll let @anovakovic01 chime in before implementing anything. I am fine with  going either way on this.

Comment (anovakovic01) [reply cmd/cassandra-reader/main.go]: @drasko purpose of enumeration is following: if you want to use logger with without environment variable, and set log level explicitly, it's better to use enum constant, than string (because compiler will give you error). In current usage this is not the case, so if you want, you can pass string for now.

**Review by anovakovic01 — CHANGES_REQUESTED** [review]

File: cmd/influxdb-writer/main.go
Comment (mfinley3) [reply cmd/influxdb-writer/main.go]: I decided that it flowed better without calling a separate function, so went with keeping these in main().

File: cmd/cassandra-reader/main.go
Comment (mfinley3) [reply cmd/cassandra-reader/main.go]: Cool, converted it back to accepting a string!

File: logger/logger_test.go
Comment (mfinley3) [reply logger/logger_test.go]: Should now all be in one for loop!

File: logger/level_test.go
Comment (anovakovic01) [inline logger/level_test.go:26]: Merge these test cases into one map and use one `for` loop.

**Review by anovakovic01 — CHANGES_REQUESTED** [review]

File: logger/level_test.go
Comment (mfinley3) [reply logger/level_test.go:26]: All cleaned up - and thanks again for the tip on adding err to the cases!

**Review by anovakovic01 — DISMISSED** [review]
@mfinley3 Thanks for your contribution! LGTM!

**Review by drasko — DISMISSED** [review]
We are there - LGTM.

Some great work here @mfinley3!

File: ws/README.md
Comment (mfinley3) [reply ws/README.md:16]: Ah shoot, you're right. I did miss it, sorry about that!

**Comment by mfinley3** [comment]
@drasko @dusanb94 @anovakovic01 
I completely forgot the mqtt-adapter. I have it working with log levels as well though. 
Thankfully the change is only a few lines since it's js and the aedeas-logging package supports it already (with an update to the latest version). 
So hopefully not too much to review!
Please take look at the updated mqtt files: https://github.com/mainflux/mainflux/pull/402/commits/89a9b80ae00fc0669dec01619376fa218cedc3c7
Sorry about this guys.

**Review by dborovcanin — APPROVED** [review]

**Review by anovakovic01 — APPROVED** [review]
LGTM!

**Comment by drasko** [comment]
Congrats @mfinley3 - you are now officially Mainflux contributor :)

**Comment by mfinley3** [comment]
@drasko And first thing I do is break the tests haha!
Looks like the fix for that is in now though.
Happy to be working with you guys a bit closer though, our team will hopefully be giving back more now. I've documented some helpful things I did during this PR for our team while working on Mainflux PRs as well.

**Pull request merged**
